src/Features/Core/Portable/EditAndContinue/EditAndContinueDiagnosticUpdateSource.cs

```diff
@@ -64,56 +64,53 @@ public void ReportDiagnostics(Workspace workspace, Solution solution, ProjectId?
-            var project = (projectId != null) ? solution.GetProject(projectId) : null;
+            var documentDiagnostics = diagnostics.WhereAsArray(d => d.DocumentId != null);
+            var projectDiagnostics = diagnostics.WhereAsArray(d => d.DocumentId == null && d.ProjectId != null);
+            var solutionDiagnostics = diagnostics.WhereAsArray(d => d.DocumentId == null && d.ProjectId == null);
 
-            foreach (var diagnostic in diagnostics)
+            if (documentDiagnostics.Length > 0)
             {
-                var document = solution.GetDocument(diagnostic.Location.SourceTree);
-
-                if (document != null)
-                {
-                    documentDiagnosticData.Add(DiagnosticData.Create(diagnostic, document));
-                }
-                else if (project != null)
+                foreach (var (documentId, diagnosticData) in documentDiagnostics.ToDictionary(data => data.DocumentId!))
                 {
-                    nonDocumentDiagnosticData.Add(DiagnosticData.Create(diagnostic, project));
-                }
-                else
-                {
-                    nonDocumentDiagnosticData.Add(DiagnosticData.Create(diagnostic, options));
+                    var diagnosticGroupId = (this, documentId, documentId.ProjectId);
+
+                    updateEvent(this, DiagnosticsUpdatedArgs.DiagnosticsCreated(
+                        diagnosticGroupId,
+                        workspace,
+                        solution,
+                        documentId.ProjectId,
+                        documentId: documentId,
+                        diagnostics: diagnosticData));
                 }
             }
 
-            if (documentDiagnosticData.Count > 0)
+            if (projectDiagnostics.Length > 0)
             {
-                foreach (var (documentId, diagnosticData) in documentDiagnosticData.ToDictionary(data => data.DocumentId))
+                foreach (var (projectId, diagnosticData) in projectDiagnostics.ToDictionary(data => data.ProjectId!))
                 {
-                    var diagnosticGroupId = (this, documentId, projectId);
+                    var diagnosticGroupId = (this, projectId);
 
                     updateEvent(this, DiagnosticsUpdatedArgs.DiagnosticsCreated(
                         diagnosticGroupId,
                         workspace,
                         solution,
                         projectId,
-                        documentId: documentId,
+                        documentId: null,
                         diagnostics: diagnosticData));
                 }
             }
 
-            if (nonDocumentDiagnosticData.Count > 0)
+            if (solutionDiagnostics.Length > 0)
             {
-                var diagnosticGroupId = (this, projectId);
+                var diagnosticGroupId = this;
 
                 updateEvent(this, DiagnosticsUpdatedArgs.DiagnosticsCreated(
                     diagnosticGroupId,
```

Comment:
for my understand, what is the group id used for?

Reply:
To group diagnostics that are reported in one event, i think.